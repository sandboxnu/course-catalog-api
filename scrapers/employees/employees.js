/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import htmlparser from "htmlparser2";
import domutils from "domutils";
import _ from "lodash";
import cookie from "cookie";
import he from "he";

import Request from "../request";
import cache from "../cache";
import macros from "../../utils/macros";
import { standardizeEmail, occurrences } from "./util";

const request = new Request("Employees");

// Scrapes from here: https://prod-web.neu.edu/wasapp/employeelookup/public/main.action

// TODO:
// Some of the phone numbers are not 10 digits. Might be able to guess area code, but it is not allways 215 just because some people have offices outside of Boston.
// Phone numbers that are not 10 digits are ignored right now.
// The detail page also has a name without the comma, but don't see a reason on scraping that instead of just splitting on comma and moving.
// Is that name ever different than the one with the comma?
// Their office is also in that detail page, but it seems like if is often different than the ones the professors have
// on their CCIS/COE/CSSH profile. Unsure if it would be worth scraping that too.

// Currently:
// Name is always scraped. This can vary slightly between different data sources.
// This name will never include accents, which the CCIS site does
// Also, multiple people have the same name so this can not be used as a primary key
// "name": "Hauck, Heather",

// Id of each employee. Always scraped.
// In Aug, 2019 this id was changed from a decimal to a base64 encoded binary string on NEU's site.
// "id": "%2B9Yas6J4nmbQ0e8J7Ju54Q%3D%3D",

// Phone number. Some people had them posted and others did not. Sometimes present, sometimes not.
// The same phone number can be listed as one person's phone number on this data source and a different person on a different data source.
// Don't have a great idea on how to handle that.
// "phone": "6173737821",

// Email. Sometimes present, sometimes not.
// "email": "h.hauck@northeastern.edu",

// Sometimes present, sometimes not. Often heavily abbreviated. (like this example)
// "primaryappointment": "Asst Dir &amp; Assoc Coop Coord",

// Always scraped.
// "primarydepartment": "DMSB Co-op"

// Should be done better here:
// employees.parseLettersResponse should be stateless, and just return some json from the html it processes
// and employees.parseLettersResponse should be sync, and not async
// and we could get rid of domutils too lmao

class Employee {
  constructor() {
    this.people = [];

    this.couldNotFindNameList = {};

    this.cookiePromise = null;
  }

  handleRequestResponce(body, callback) {
    const handler = new htmlparser.DomHandler(callback);
    const parser = new htmlparser.Parser(handler);
    parser.write(body);
    parser.done();
  }

  //returns a {colName:[values]} where colname is the first in the column
  //regardless if its part of the header or the first row of the body
  parseTable(table) {
    if (table.name !== "table") {
      macros.error("parse table was not given a table..");
      return null;
    }

    //includes both header rows and body rows
    const rows = domutils.getElementsByTagName("tr", table);

    if (rows.length === 0) {
      return null;
    }

    const retVal = {};
    const heads = [];

    //the headers
    rows[0].children.forEach((element) => {
      if (element.type !== "tag" || ["th", "td"].indexOf(element.name) === -1) {
        return;
      }

      const text = domutils
        .getText(element)
        .trim()
        .toLowerCase()
        .replace(/\s/gi, "");
      retVal[text] = [];
      heads.push(text);
    });

    //add the other rows
    rows.slice(1).forEach((row) => {
      let index = 0;
      row.children.forEach((element) => {
        if (
          element.type !== "tag" ||
          ["th", "td"].indexOf(element.name) === -1
        ) {
          return;
        }
        if (index >= heads.length) {
          macros.log(
            "warning, table row is longer than head, ignoring content",
            index,
            heads,
            rows
          );
          return;
        }

        retVal[heads[index]].push(domutils.getText(element).trim());

        //only count valid elements, not all row.children
        index++;
      });

      //add empty strings until reached heads length
      for (; index < heads.length; index++) {
        retVal[heads[index]].push("");
      }
    });
    return {
      rowCount: rows.length - 1,
      parsedTable: retVal,
    };
  }

  async getCookiePromise() {
    if (this.cookiePromise) {
      return this.cookiePromise;
    }

    macros.verbose("neu employee getting cookie");

    this.cookiePromise = request
      .get({
        url:
          "https://prod-web.neu.edu/wasapp/employeelookup/public/main.action",
      })
      .then((resp) => {
        // Parse the cookie from the response
        const cookieString = resp.headers["set-cookie"][0];
        const cookies = cookie.parse(cookieString);
        return cookies.JSESSIONID;
      });

    return this.cookiePromise;
  }

  hitWithLetters(lastNameStart, jsessionCookie) {
    return request.get({
      url: `https://prod-web.neu.edu/wasapp/employeelookup/public/searchEmployees.action?searchBy=Last+Name&queryType=begins+with&searchText=${lastNameStart}&deptText=&addrText=&numText=&divText=&facStaff=2`,
      headers: {
        Cookie: `JSESSIONID=${jsessionCookie}`,
      },
    });
  }

  // Given a list of things, will find the first one that is longer than 1 letter (a-z)
  findName(list) {
    for (let i = 0; i < list.length; i++) {
      const noSymbols = list[i].toLowerCase().replace(/[^0-9a-zA-Z]/gi, "");

      if (
        noSymbols.length > 1 &&
        !["ii", "iii", "jr", "sr", "dr"].includes(noSymbols)
      ) {
        return list[i];
      }
    }

    // Only log each warning once, just to not spam the macros. This method is called a lot.
    const logMatchString = list.join("");
    if (this.couldNotFindNameList[logMatchString]) {
      return null;
    }
    this.couldNotFindNameList[logMatchString] = true;

    macros.log("Could not find name from list:", list);
    return null;
  }

  // This splits the employee name by the comma in the middle and the name after the comma is the first name and the name before is the last name
  // The util function for this does not split the name by the comma, it assumes that the name is just separated by spaces.
  getFirstLastName(name) {
    if (name.match(/jr.?,/gi)) {
      name = name.replace(/, jr.?,/gi, ",");
    }

    name = _.trim(name, ",");

    if (occurrences(name, ",") !== 1) {
      macros.log("Name has != 1 commas", name);
      return null;
    }

    const splitOnComma = name.split(",");

    const beforeCommaSplit = splitOnComma[1].trim().split(" ");
    const firstName = this.findName(beforeCommaSplit);

    const afterCommaSplit = splitOnComma[0].trim().split(" ").reverse();
    const lastName = this.findName(afterCommaSplit);

    const retVal = {};
    retVal.firstName = firstName;
    retVal.lastName = lastName;
    return retVal;
  }

  parseLettersResponse(response, lastNameStart) {
    return new Promise((resolve, reject) => {
      this.handleRequestResponce(response.body, (err, dom) => {
        const elements = domutils.getElementsByTagName("table", dom);

        for (let i = 0; i < elements.length; i++) {
          const element = elements[i];

          const goal = {
            width: "100%",
          };

          if (_.isEqual(element.attribs, goal)) {
            // Delete one of the elements that is before the header that would mess stuff up
            domutils.removeElement(element.children[1].children[1]);

            const tableData = this.parseTable(element);

            if (!tableData) {
              return resolve();
            }

            const parsedTable = tableData.parsedTable;
            const rowCount = tableData.rowCount;

            macros.log("Found", rowCount, " people on page ", lastNameStart);

            for (let j = 0; j < rowCount; j++) {
              const person = {};
              const nameWithComma = he
                .decode(parsedTable.name[j])
                .split("\n\n")[0];

              if (nameWithComma.includes("Do Not Use ")) {
                macros.log("Skipping entry that says Do Not Use.");
                continue;
              }

              // Put the first name before the last name.
              // Another way to do this would be to hit the detail section of each employee and scrape the name from the title.
              const commaNameSplit = nameWithComma.split(",");

              for (let k = 0; k < commaNameSplit.length; k++) {
                commaNameSplit[k] = commaNameSplit[k].trim();
              }

              // Remove Jr and Jr.
              _.pull(commaNameSplit, "Jr", "Jr.");

              if (commaNameSplit.length > 2) {
                macros.warn(
                  "Has more than one comma skipping.",
                  commaNameSplit
                );
                person.name = nameWithComma;
              } else {
                person.name = `${commaNameSplit[1].trim()} ${commaNameSplit[0].trim()}`;
              }

              // Generate first name and last name from the name on the person
              const firstLastName = this.getFirstLastName(nameWithComma);
              if (!firstLastName) {
                continue;
              }
              const { firstName, lastName } = firstLastName;
              if (firstName && lastName) {
                person.firstName = firstName;
                person.lastName = lastName;
              }

              // The "#anchor_68409676" tags on the page are randomly generated each time the page is loaded,
              // so use this hrefparameter id instead, which should be more stable.
              // This hrefparameter param is a url encoded, base64 string.
              // We could decode it, but there's just no point.
              const idMatch = parsedTable.name[j].match(
                /.hrefparameter\s+=\s+"id=([\d\w%+/]+)";/i
              );
              if (!idMatch) {
                macros.warn(
                  "Unable to parse id, using random number",
                  nameWithComma
                );
                person.id = String(Math.random());
              } else {
                let id = idMatch[1];

                // Trim the id if it is too long. This is just a sanity check.
                // 35 is just a arbitrary decided number.
                if (id.length > 35) {
                  macros.warn("person id over 35 chars?", id);
                  id = id.slice(0, 35);
                }

                person.id = id;
              }

              let phone = parsedTable.phone[j];
              phone = phone.replace(/\D/g, "");

              // Maybe add support for guesing area code if it is ommitted and most of the other ones have the same area code
              if (phone.length === 10) {
                person.phone = phone;
              }

              // Scrape the email from the table
              const email = standardizeEmail(parsedTable.email[j]);
              if (email) {
                person.emails = [email];
              }

              // Scrape the primaryappointment
              const primaryappointment = parsedTable.primaryappointment[j];
              if (
                primaryappointment &&
                primaryappointment !== "Not Available"
              ) {
                person.primaryRole = he.decode(
                  parsedTable.primaryappointment[j]
                );
              }

              // Scrape the primary department
              person.primaryDepartment = he.decode(
                parsedTable.primarydepartment[j]
              );

              // Add it to the people list
              this.people.push(person);
            }
            return resolve();
          }
        }

        macros.error("YOOOOO it didnt find the table", response.body.length);
        macros.error(response.body);
        return reject();
      });
    });
  }

  async get(lastNameStart) {
    const jsessionCookie = await this.getCookiePromise();

    macros.verbose("neu employee got cookie", jsessionCookie);

    const response = await this.hitWithLetters(lastNameStart, jsessionCookie);

    return this.parseLettersResponse(response, lastNameStart);
  }

  async main() {
    // if this is dev and this data is already scraped, just return the data
    if (macros.DEV && require.main !== module) {
      const devData = await cache.get(
        macros.DEV_DATA_DIR,
        this.constructor.name,
        "main"
      );
      if (devData) {
        return devData;
      }
    }

    const promises = [];

    const alphabetArray = macros.ALPHABET.split("");

    for (const firstLetter of alphabetArray) {
      for (const secondLetter of alphabetArray) {
        promises.push(this.get(firstLetter + secondLetter));
      }
    }

    await Promise.all(promises);

    macros.verbose("Done all employee requests");

    if (macros.DEV) {
      await cache.set(
        macros.DEV_DATA_DIR,
        this.constructor.name,
        "main",
        this.people
      );
      macros.log(this.people.length, "employees saved to a file!");
    }

    return this.people;
  }
}

const instance = new Employee();

export default instance;

if (require.main === module) {
  instance.main();
}
