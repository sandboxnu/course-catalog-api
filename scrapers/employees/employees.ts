/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import Request from "../request";
import cache from "../cache";
import macros from "../../utils/macros";
import { Employee, EmployeeRequestResponse } from "../../types/types";
import { v4 as uuidv4 } from "uuid";

const request = new Request("Employees");

class NeuEmployee {
  people: Employee[];

  constructor() {
    this.people = [];
  }

  /**
   * The NEU API returns 'Not Available' for some fields instead of making them null.
   * Given an optional string, return null if the string is undefined or 'Not Available'
   */
  private parseOptionalParameter(parameter?: string): string | null {
    if (!parameter || parameter.toLocaleLowerCase() === "not available") {
      return null;
    }
    return parameter;
  }

  /**
   * Given the raw data for one employee from NEU's API, convert it to our
   * expected Employee type.
   */
  private parseEmployee(employee: EmployeeRequestResponse): Employee {
    const email = this.parseOptionalParameter(employee.Email);
    const primaryRole = this.parseOptionalParameter(employee.PositionTitle);
    const longPrimaryRole = this.parseOptionalParameter(
      employee.LongPositionTitle,
    );

    return {
      id: email ?? uuidv4(),
      name: `${employee.FirstName} ${employee.LastName}`,
      firstName: employee.FirstName,
      lastName: employee.LastName,
      primaryDepartment: employee.Department,
      primaryRole: primaryRole ?? longPrimaryRole,
      phone: this.parseOptionalParameter(employee.PhoneNumber),
      email: email,
      officeRoom: this.parseOptionalParameter(employee.CampusAddress),
    };
  }

  parseApiResponse(response: EmployeeRequestResponse[]): Employee[] {
    return (
      response
        .map((employee) => this.parseEmployee(employee))
        // Northeastern likes testing in prod (don't we all)
        // As of 2022-03, they have 4 employees whose names are variations of "do not use"
        //    Do Not Use, Ed
        //    Do Not Use Hrm See 000499418, Do Not Use
        //    DO NOT USE, DO NOT USE
        .filter(
          (employee) => !employee.name.toLowerCase().includes("do not use"),
        )
    );
  }

  // Scrapes the source code to get the X-CSRF token.
  async queryXcsrfToken(): Promise<string> {
    return request
      .get("https://nu.outsystemsenterprise.com/FSD/scripts/OutSystems.js")
      .then((resp) => {
        return resp.body.match(/"X-CSRFToken".*?="(.*?)"/i)[1];
      });
  }

  // Queries the API module version from the API (needed for requests)
  async queryModuleVersion(): Promise<string> {
    return request
      .get(
        "https://nu.outsystemsenterprise.com/FSD/moduleservices/moduleversioninfo",
      )
      .then((resp) => JSON.parse(resp.body)["versionToken"]);
  }

  /**
   * Queries the Northeastern employees API.
   */
  async queryEmployeesApi(): Promise<void> {
    // These values are taken from the API exposed here: https://nu.outsystemsenterprise.com/FSD/
    // The purpose of these values is unknown, as the API is undocumented. If requests start failing, please
    // inspect the API response at the above link & update accordingly
    const csrfToken = await this.queryXcsrfToken();
    const moduleVersion = await this.queryModuleVersion();

    const employeeQuery = {
      versionInfo: {
        moduleVersion,
        apiVersion: "Zox0cGCdj_7M+WZErCpkhQ",
      },
      viewName: "MainFlow.FacultyAndStaffDirectory",
      // Their frontend requires non-null queries, but the backend doesn't :)
      // If this changes at any point, we used to do a cartesian product of the alphabet
      // ('aa', 'ab', etc.). See https://github.com/sandboxnu/course-catalog-api/pull/97/files for more
      inputParameters: {
        Input_Name1: "",
        Input_Name2: "",
      },
    };

    const response = await request.post(
      "https://nu.outsystemsenterprise.com/FSD/screenservices/FSD/MainFlow/Name/ActionGetContactsByName_NonSpecificType",
      {
        json: employeeQuery,
        headers: {
          "X-CSRFToken": csrfToken,
        },
      },
    );

    const parsedResponse = JSON.parse(response.body);
    const employees: EmployeeRequestResponse[] =
      parsedResponse?.data?.EmployeeDirectoryContact?.List;

    this.people = this.parseApiResponse(employees);
  }

  async main(): Promise<Employee[]> {
    // if this is dev and this data is already scraped, just return the data
    if (macros.DEV && require.main !== module) {
      const devData = await cache.get(
        macros.DEV_DATA_DIR,
        this.constructor.name,
        "main",
      );
      if (devData) {
        return (devData as Employee[]).map((employee) => {
          const id = employee?.email ?? uuidv4();
          return { ...employee, id };
        });
      }
    }

    await this.queryEmployeesApi();

    macros.verbose("Done all employee requests");

    if (macros.DEV) {
      await cache.set(
        macros.DEV_DATA_DIR,
        this.constructor.name,
        "main",
        this.people,
      );
      macros.log(this.people.length, "NEU employees saved.");
    }

    return this.people;
  }
}

const instance = new NeuEmployee();

export default instance;

if (require.main === module) {
  instance.main();
}
