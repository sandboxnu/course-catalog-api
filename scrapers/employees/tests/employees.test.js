/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import employeeData from "./data/employees/employee_results_BE.json";

import employees from "../employees";
import { validate } from "uuid";

// Test to make sure parsing of an employees result page stays the same
it("should be able to parse API responses", async () => {
  employees.parseLettersResponse(employeeData);

  // As documented in the employees.ts file, the decoded string should be 16 bytes.
  // Idk what those bytes mean, but we can check the length
  for (const employee of employees.people) {
    // https://stackoverflow.com/questions/46155/whats-the-best-way-to-validate-an-email-address-in-javascript
    const validId = employee.id.match(/\S+@\S+\.\S+/) || validate(employee.id);
    expect(validId).toBeTruthy();
  }

  expect(employees).toMatchSnapshot();
});
