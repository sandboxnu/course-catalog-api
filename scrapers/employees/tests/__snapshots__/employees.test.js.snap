// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be able to parse API responses 1`] = `
NeuEmployee {
  "couldNotFindNameList": Object {},
  "csrfToken": null,
  "people": Array [
    Object {
      "email": "t.abbas@northeastern.edu",
      "emails": Array [
        "t.abbas@northeastern.edu",
      ],
      "firstName": "Tabeer",
      "id": "t.abbas@northeastern.edu",
      "lastName": "Abbas",
      "name": "Tabeer Abbas",
      "office": "Not Available",
      "officeRoom": "Not Available",
      "phone": "Not Available",
      "primaryDepartment": "College of Professional Studies",
      "primaryRole": "Lecturer",
    },
    Object {
      "email": "m.abedi@northeastern.edu",
      "emails": Array [
        "m.abedi@northeastern.edu",
      ],
      "firstName": "Mehdi",
      "id": "m.abedi@northeastern.edu",
      "lastName": "Abedi",
      "name": "Mehdi Abedi",
      "office": "334 SN",
      "officeRoom": "334 SN",
      "phone": "6173734427",
      "primaryDepartment": "Mechanical + Industrial Engineering",
      "primaryRole": "Associate Teaching Professor",
    },
    Object {
      "email": "h.abeels@northeastern.edu",
      "emails": Array [
        "h.abeels@northeastern.edu",
      ],
      "firstName": "Hannah",
      "id": "h.abeels@northeastern.edu",
      "lastName": "Abeels",
      "name": "Hannah Abeels",
      "office": "KN 120 Knowles",
      "officeRoom": "KN 120 Knowles",
      "phone": "6173735149",
      "primaryDepartment": "School Of Law-Academic",
      "primaryRole": "Asst Dir - Online & Hybrid Pro",
    },
    Object {
      "email": "s.abel@northeastern.edu",
      "emails": Array [
        "s.abel@northeastern.edu",
      ],
      "firstName": "Sherry",
      "id": "s.abel@northeastern.edu",
      "lastName": "Abel",
      "name": "Sherry Abel",
      "office": "101 N. Tyron St",
      "officeRoom": "101 N. Tyron St",
      "phone": "9802248466",
      "primaryDepartment": "School Of Nursing",
      "primaryRole": "Part-time Lecturer",
    },
    Object {
      "email": "p.abelli@northeastern.edu",
      "emails": Array [
        "p.abelli@northeastern.edu",
      ],
      "firstName": "Peter",
      "id": "p.abelli@northeastern.edu",
      "lastName": "Abelli",
      "name": "Peter Abelli",
      "office": "140 CN",
      "officeRoom": "140 CN",
      "phone": "6173735599",
      "primaryDepartment": "Maintenance-HVAC-Dorms",
      "primaryRole": "HVAC Systems Technician",
    },
    Object {
      "email": "m.abels@northeastern.edu",
      "emails": Array [
        "m.abels@northeastern.edu",
      ],
      "firstName": "Margot",
      "id": "m.abels@northeastern.edu",
      "lastName": "Abels",
      "name": "Margot Abels",
      "office": "310 RP",
      "officeRoom": "310 RP",
      "phone": "Not Available",
      "primaryDepartment": "Human Services",
      "primaryRole": "Assistant Teaching Professor",
    },
    Object {
      "email": "e.ackerman@northeastern.edu",
      "emails": Array [
        "e.ackerman@northeastern.edu",
      ],
      "firstName": "Elizabeth",
      "id": "e.ackerman@northeastern.edu",
      "lastName": "Ackerman",
      "name": "Elizabeth Ackerman",
      "office": "102 RB",
      "officeRoom": "102 RB",
      "phone": "Not Available",
      "primaryDepartment": "BCHS Office of the Dean",
      "primaryRole": "Part-Time Lecturer",
    },
    Object {
      "email": "m.acosta@northeastern.edu",
      "emails": Array [
        "m.acosta@northeastern.edu",
      ],
      "firstName": "Beth",
      "id": "m.acosta@northeastern.edu",
      "lastName": "Acosta",
      "name": "Beth Acosta",
      "office": "RB 103 RB",
      "officeRoom": "RB 103 RB",
      "phone": "6173735177",
      "primaryDepartment": "School Of Nursing",
      "primaryRole": "Part-time Lecturer",
    },
    Object {
      "email": "m.adams@northeastern.edu",
      "emails": Array [
        "m.adams@northeastern.edu",
      ],
      "firstName": "Marybeth",
      "id": "m.adams@northeastern.edu",
      "lastName": "Adams-Khoury",
      "name": "Marybeth Adams-Khoury",
      "office": "102 RB",
      "officeRoom": "102 RB",
      "phone": "Not Available",
      "primaryDepartment": "BCHS Office of the Dean",
      "primaryRole": "Part-Time Lecturer",
    },
    Object {
      "email": "j.adelberg@northeastern.edu",
      "emails": Array [
        "j.adelberg@northeastern.edu",
      ],
      "firstName": "Jeff",
      "id": "j.adelberg@northeastern.edu",
      "lastName": "Adelberg",
      "name": "Jeff Adelberg",
      "office": "360 Huntington Ave",
      "officeRoom": "360 Huntington Ave",
      "phone": "Not Available",
      "primaryDepartment": "Theater",
      "primaryRole": "Part-Time Lecturer",
    },
    Object {
      "email": "k.ahkee@northeastern.edu",
      "emails": Array [
        "k.ahkee@northeastern.edu",
      ],
      "firstName": "Isabel",
      "id": "k.ahkee@northeastern.edu",
      "lastName": "Ah Kee",
      "name": "Isabel Ah Kee",
      "office": "122 SS",
      "officeRoom": "122 SS",
      "phone": "6173737745",
      "primaryDepartment": "Audit & Advisory Services",
      "primaryRole": "Dir - Audit",
    },
    Object {
      "email": "b.ahmadi@northeastern.edu",
      "emails": Array [
        "b.ahmadi@northeastern.edu",
      ],
      "firstName": "Behzad",
      "id": "b.ahmadi@northeastern.edu",
      "lastName": "Ahmadi",
      "name": "Behzad Ahmadi",
      "office": "Not Available",
      "officeRoom": "Not Available",
      "phone": "Not Available",
      "primaryDepartment": "College of Professional Studies",
      "primaryRole": "Lecturer",
    },
    Object {
      "email": "t.roberts@northeastern.edu",
      "emails": Array [
        "t.roberts@northeastern.edu",
      ],
      "firstName": "Therese",
      "id": "t.roberts@northeastern.edu",
      "lastName": "Ajtum-Roberts",
      "name": "Therese Ajtum-Roberts",
      "office": "Not Available",
      "officeRoom": "Not Available",
      "phone": "Not Available",
      "primaryDepartment": "College of Professional Studies",
      "primaryRole": "Lecturer",
    },
    Object {
      "email": "j.aliberti@northeastern.edu",
      "emails": Array [
        "j.aliberti@northeastern.edu",
      ],
      "firstName": "Julie",
      "id": "j.aliberti@northeastern.edu",
      "lastName": "Aliberti",
      "name": "Julie Aliberti",
      "office": "360 Huntington Ave",
      "officeRoom": "360 Huntington Ave",
      "phone": "Not Available",
      "primaryDepartment": "Financial Aid",
      "primaryRole": "Sr Asst Dir - Financial Counse",
    },
    Object {
      "email": "e.allen@northeastern.edu",
      "emails": Array [
        "e.allen@northeastern.edu",
      ],
      "firstName": "Elizabeth",
      "id": "e.allen@northeastern.edu",
      "lastName": "Allen",
      "name": "Elizabeth Allen",
      "office": "310 RP",
      "officeRoom": "310 RP",
      "phone": "6173738122",
      "primaryDepartment": "Provost Research",
      "primaryRole": "Research Development Officer",
    },
    Object {
      "email": "b.almand-hunter@northeastern.edu",
      "emails": Array [
        "b.almand-hunter@northeastern.edu",
      ],
      "firstName": "Berkeley",
      "id": "b.almand-hunter@northeastern.edu",
      "lastName": "Almand-Hunter",
      "name": "Berkeley Almand-Hunter",
      "office": "100 Commercial St., Suite 400",
      "officeRoom": "100 Commercial St., Suite 400",
      "phone": "Not Available",
      "primaryDepartment": "College of Professional Studies",
      "primaryRole": "Lecturer",
    },
    Object {
      "email": "r.amini@northeastern.edu",
      "emails": Array [
        "r.amini@northeastern.edu",
      ],
      "firstName": "Rouzbeh",
      "id": "r.amini@northeastern.edu",
      "lastName": "Amini",
      "name": "Rouzbeh Amini",
      "office": "334 SN",
      "officeRoom": "334 SN",
      "phone": "6173737688",
      "primaryDepartment": "Mechanical + Industrial Engineering",
      "primaryRole": "Associate Professor",
    },
    Object {
      "email": "r.anzuoni@northeastern.edu",
      "emails": Array [
        "r.anzuoni@northeastern.edu",
      ],
      "firstName": "Rebecca",
      "id": "r.anzuoni@northeastern.edu",
      "lastName": "Anzuoni",
      "name": "Rebecca Anzuoni",
      "office": "598 CP",
      "officeRoom": "598 CP",
      "phone": "6173737780",
      "primaryDepartment": "University Marketing",
      "primaryRole": "Vice President - Marketing",
    },
    Object {
      "email": "e.arone@northeastern.edu",
      "emails": Array [
        "e.arone@northeastern.edu",
      ],
      "firstName": "Elizabeth",
      "id": "e.arone@northeastern.edu",
      "lastName": "Arone",
      "name": "Elizabeth Arone",
      "office": "360 Huntington Ave",
      "officeRoom": "360 Huntington Ave",
      "phone": "Not Available",
      "primaryDepartment": "Health Sciences",
      "primaryRole": "Administrative Coordinator",
    },
    Object {
      "email": "b.ashburn@northeastern.edu",
      "emails": Array [
        "b.ashburn@northeastern.edu",
      ],
      "firstName": "Ben",
      "id": "b.ashburn@northeastern.edu",
      "lastName": "Ashburn",
      "name": "Ben Ashburn",
      "office": "360 Huntington Avenue",
      "officeRoom": "360 Huntington Avenue",
      "phone": "Not Available",
      "primaryDepartment": "College of Professional Studies",
      "primaryRole": "Lecturer",
    },
    Object {
      "email": "ro.austin@northeastern.edu",
      "emails": Array [
        "ro.austin@northeastern.edu",
      ],
      "firstName": "Robert",
      "id": "ro.austin@northeastern.edu",
      "lastName": "Austin",
      "name": "Robert Austin",
      "office": "01 GG",
      "officeRoom": "01 GG",
      "phone": "6177992748",
      "primaryDepartment": "Facilities Administration-10",
      "primaryRole": "Assoc Dir - Customer Service",
    },
    Object {
      "email": "i.bachman@northeastern.edu",
      "emails": Array [
        "i.bachman@northeastern.edu",
      ],
      "firstName": "Isabella",
      "id": "i.bachman@northeastern.edu",
      "lastName": "Bachman",
      "name": "Isabella Bachman",
      "office": "HA",
      "officeRoom": "HA",
      "phone": "6173734108",
      "primaryDepartment": "DMSB Communications",
      "primaryRole": "Communications Specialist",
    },
    Object {
      "email": "r.baginski@northeastern.edu",
      "emails": Array [
        "r.baginski@northeastern.edu",
      ],
      "firstName": "Robert",
      "id": "r.baginski@northeastern.edu",
      "lastName": "Baginski",
      "name": "Robert Baginski",
      "office": "RB",
      "officeRoom": "RB",
      "phone": "6173733195",
      "primaryDepartment": "Physician Assistant Program",
      "primaryRole": "Associate Clinical Professor",
    },
    Object {
      "email": "b.ballaro@northeastern.edu",
      "emails": Array [
        "b.ballaro@northeastern.edu",
      ],
      "firstName": "Beverly",
      "id": "b.ballaro@northeastern.edu",
      "lastName": "Ballaro",
      "name": "Beverly Ballaro",
      "office": "50 NI",
      "officeRoom": "50 NI",
      "phone": "Not Available",
      "primaryDepartment": "College of Professional Studies",
      "primaryRole": "Sr Student Affairs & Special P",
    },
    Object {
      "email": "l.bamberger@northeastern.edu",
      "emails": Array [
        "l.bamberger@northeastern.edu",
      ],
      "firstName": "Leah",
      "id": "l.bamberger@northeastern.edu",
      "lastName": "Bamberger",
      "name": "Leah Bamberger",
      "office": "360 Huntington Avenue",
      "officeRoom": "360 Huntington Avenue",
      "phone": "Not Available",
      "primaryDepartment": "Facilities Administration-10",
      "primaryRole": "Exec Dir - Climate Justice & S",
    },
    Object {
      "email": "b.bamgbade@northeastern.edu",
      "emails": Array [
        "b.bamgbade@northeastern.edu",
      ],
      "firstName": "Benita",
      "id": "b.bamgbade@northeastern.edu",
      "lastName": "Bamgbade",
      "name": "Benita Bamgbade",
      "office": "218 TF",
      "officeRoom": "218 TF",
      "phone": "6173737973",
      "primaryDepartment": "Pharm & Hlth Systems Sciences",
      "primaryRole": "Assistant Professor",
    },
    Object {
      "email": "a.barabasi@northeastern.edu",
      "emails": Array [
        "a.barabasi@northeastern.edu",
      ],
      "firstName": "Albert-Laszlo",
      "id": "a.barabasi@northeastern.edu",
      "lastName": "Barabasi",
      "name": "Albert-Laszlo Barabasi",
      "office": "1100 - 177",
      "officeRoom": "1100 - 177",
      "phone": "6173737774",
      "primaryDepartment": "Physics",
      "primaryRole": "Dodge Prof NetSI & Dist Prof",
    },
    Object {
      "email": "a.barber@northeastern.edu",
      "emails": Array [
        "a.barber@northeastern.edu",
      ],
      "firstName": "Doctor Arthur",
      "id": "a.barber@northeastern.edu",
      "lastName": "Barber",
      "name": "Doctor Arthur Barber",
      "office": "360 Huntington Ave",
      "officeRoom": "360 Huntington Ave",
      "phone": "Not Available",
      "primaryDepartment": "College of Professional Studies",
      "primaryRole": "Lecturer",
    },
    Object {
      "email": "ju.barber@northeastern.edu",
      "emails": Array [
        "ju.barber@northeastern.edu",
      ],
      "firstName": "Julie",
      "id": "ju.barber@northeastern.edu",
      "lastName": "Barber",
      "name": "Julie Barber",
      "office": "Not Available",
      "officeRoom": "Not Available",
      "phone": "Not Available",
      "primaryDepartment": "College of Professional Studies",
      "primaryRole": "Lecturer",
    },
    Object {
      "email": "p.barber@northeastern.edu",
      "emails": Array [
        "p.barber@northeastern.edu",
      ],
      "firstName": "Paul",
      "id": "p.barber@northeastern.edu",
      "lastName": "Barber",
      "name": "Paul Barber",
      "office": "100 CP",
      "officeRoom": "100 CP",
      "phone": "Not Available",
      "primaryDepartment": "Public Safety",
      "primaryRole": "Security Technician",
    },
    Object {
      "email": "e.barberis@northeastern.edu",
      "emails": Array [
        "e.barberis@northeastern.edu",
      ],
      "firstName": "Emanuela",
      "id": "e.barberis@northeastern.edu",
      "lastName": "Barberis",
      "name": "Emanuela Barberis",
      "office": "111 DA",
      "officeRoom": "111 DA",
      "phone": "6173732935",
      "primaryDepartment": "Physics",
      "primaryRole": "Professor",
    },
    Object {
      "email": "s.barenberg@northeastern.edu",
      "emails": Array [
        "s.barenberg@northeastern.edu",
      ],
      "firstName": "Sumner",
      "id": "s.barenberg@northeastern.edu",
      "lastName": "Barenberg",
      "name": "Sumner Barenberg",
      "office": "129 ISEC",
      "officeRoom": "129 ISEC",
      "phone": "6173737805",
      "primaryDepartment": "Bioengineering",
      "primaryRole": "Professor of the Practice",
    },
  ],
}
`;
